File: clients/java/Ttid.java
// TTID client — drives the `ttid` binary's persistent NDJSON loop.
//
// No dependencies (java.lang.Process only). Requires the `ttid` binary on PATH
// or an explicit path. One long-lived subprocess.
//
//   try (Ttid t = new Ttid()) {
//       String id = t.generate();          // {"...","result":"4VL..."}
//       String up = t.generate("4VL...");  // advance it
//       String times = t.decodeTime("4VL...");
//       String valid = t.isTTID("4VL...");
//       String uuid = t.isUUID("...");
//   }
//
// Each method builds the request, checks it succeeded, and returns the raw JSON
// response line (parse `result` with Jackson/Gson). Method names follow Java's
// camelCase. request(json) is the raw escape hatch.

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.List;

public final class Ttid implements AutoCloseable {
    private final Process proc;
    private final BufferedWriter in;
    private final BufferedReader out;

    public Ttid() throws IOException {
        this("ttid");
    }

    public Ttid(String binary) throws IOException {
        this.proc = new ProcessBuilder(List.of(binary, "exec", "--loop"))
                .redirectError(ProcessBuilder.Redirect.INHERIT)
                .start();
        this.in = new BufferedWriter(
                new OutputStreamWriter(proc.getOutputStream(), StandardCharsets.UTF_8));
        this.out = new BufferedReader(
                new InputStreamReader(proc.getInputStream(), StandardCharsets.UTF_8));
    }

    /** Send one raw machine-protocol op (JSON string); returns the response line. */
    public synchronized String request(String opJson) throws IOException {
        if (!proc.isAlive()) throw new IOException("ttid process has exited");
        in.write(opJson.stripTrailing());
        in.write('\n');
        in.flush();
        String line = out.readLine();
        if (line == null) throw new IOException("ttid closed the stream");
        return line;
    }

    // Build an op from native fields (skipping null values), send it, and error
    // on a failure response. ponytail: checks the "ok":true field by substring.
    private String op(String name, Object... kv) throws IOException {
        StringBuilder sb = new StringBuilder("{\"op\":").append(toJson(name));
        for (int i = 0; i + 1 < kv.length; i += 2) {
            if (kv[i + 1] == null) continue;
            sb.append(',').append(toJson(kv[i].toString())).append(':').append(toJson(kv[i + 1]));
        }
        String resp = request(sb.append('}').toString());
        if (!resp.contains("\"ok\":true")) throw new IOException(resp.strip());
        return resp;
    }

    /** Generate a new TTID (id == null), or advance it (del == true to tombstone). */
    public String generate() throws IOException {
        return op("generate");
    }

    public String generate(String id) throws IOException {
        return op("generate", "id", id);
    }

    public String generate(String id, boolean del) throws IOException {
        return op("generate", "id", id, "delete", del ? Boolean.TRUE : null);
    }

    public String decodeTime(String id) throws IOException {
        return op("decodeTime", "id", id);
    }

    public String isTTID(String id) throws IOException {
        return op("isTTID", "id", id);
    }

    public String isUUID(String id) throws IOException {
        return op("isUUID", "id", id);
    }

    // Minimal JSON encoder for String / Number / Boolean / null.
    static String toJson(Object v) {
        if (v == null) return "null";
        if (v instanceof String) {
            return "\"" + ((String) v).replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
        }
        if (v instanceof Boolean || v instanceof Number) return v.toString();
        throw new IllegalArgumentException("unsupported JSON value: " + v.getClass());
    }

    @Override
    public void close() throws IOException {
        in.close(); // EOF ends the loop
        try {
            proc.waitFor();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        out.close();
    }
}

File: clients/kotlin/Ttid.kt
// TTID client — drives the `ttid` binary's persistent NDJSON loop.
//
// No dependencies (java.lang.ProcessBuilder only). Requires the `ttid` binary on
// PATH or an explicit path. One long-lived subprocess.
//
//   Ttid().use { t ->
//       val id = t.generate()        // {..."result":"4VL..."}
//       t.generate("4VL...")         // advance it
//       t.generate("4VL...", true)   // mark deleted
//       val times = t.decodeTime("4VL...") // methods return the raw JSON
//       val valid = t.isTTID("4VL...")     // response line — parse with kotlinx/Gson
//   }
//
// Each method builds the request, checks it succeeded, and returns the raw JSON
// response line. Method names follow Kotlin's camelCase. request(json) is the
// raw escape hatch.

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

class Ttid(binary: String = "ttid") : AutoCloseable {
    private val proc = ProcessBuilder(binary, "exec", "--loop")
        .redirectError(ProcessBuilder.Redirect.INHERIT)
        .start()
    private val input = BufferedWriter(OutputStreamWriter(proc.outputStream, Charsets.UTF_8))
    private val output = BufferedReader(InputStreamReader(proc.inputStream, Charsets.UTF_8))

    /** Send one raw machine-protocol op (JSON string); returns the response line. */
    @Synchronized
    fun request(opJson: String): String {
        if (!proc.isAlive) throw RuntimeException("ttid process has exited")
        input.write(opJson.trimEnd())
        input.write("\n")
        input.flush()
        return output.readLine() ?: throw RuntimeException("ttid closed the stream")
    }

    // Build an op from native fields (skipping null values), send it, and error
    // on a failure response. ponytail: checks the "ok":true field by substring.
    private fun op(name: String, vararg kv: Pair<String, Any?>): String {
        val sb = StringBuilder("{\"op\":").append(toJson(name))
        for ((key, value) in kv) {
            if (value == null) continue
            sb.append(',').append(toJson(key)).append(':').append(toJson(value))
        }
        val resp = request(sb.append('}').toString())
        if (!resp.contains("\"ok\":true")) throw RuntimeException(resp.trim())
        return resp
    }

    /** Generate a new TTID (id == null), or advance it (delete == true to tombstone). */
    fun generate(id: String? = null, delete: Boolean = false): String =
        op("generate", "id" to id, "delete" to if (delete) true else null)

    fun decodeTime(id: String): String = op("decodeTime", "id" to id)
    fun isTTID(id: String): String = op("isTTID", "id" to id)
    fun isUUID(id: String): String = op("isUUID", "id" to id)

    override fun close() {
        input.close() // EOF ends the loop
        proc.waitFor()
        output.close()
    }

    companion object {
        // Minimal JSON encoder for String / Number / Boolean / null.
        fun toJson(v: Any?): String = when (v) {
            null -> "null"
            is String -> "\"" + v.replace("\\", "\\\\").replace("\"", "\\\"") + "\""
            is Boolean, is Number -> v.toString()
            else -> throw IllegalArgumentException("unsupported JSON value: $v")
        }
    }
}
